### PR TITLE
clusterstorage first pass

### PR DESCRIPTION
**Pull request by jr0d**
# D2IQ ticket
https://jira.d2iq.com/browse/COPS-6792

# tl;dr
This change is required to overcome the browser storage size limitations. Specifically, group claims exceeding 4k in size result gorilla session creation failures. To solve for this tfa can be configured to store session data in kubernetes secrets. 

# Kubernetes Cluster Storage

This PR adds the capability to store userInfo claim data using the kubernetes apiserver. The changes contain:

* client library operations
* an unbound TTL cache implementation
* a simple garbage collector for removing stale records 

## Details

On successful authentication callback, id token claims (email, username, groups) are extracted and stored in a UserInfo structure. A claimId is generated and stored in the users browser. Subsequent requests containing the claimId is authenticated by a sha256 hmac. The UserInfo is serialized to JSON and stored in a kubernetes secret within a user defined namespace. Secrets are referenced via the following labels:

```
        ClaimsLabel    = "traefik-forward-auth/claim"
	ClaimsIDLabel  = "traefik-forward-auth/claim-id"
```

When a user makes a request using an authenticated claimId, this userinfo fetched from the corresponding secret on the API server. In order to reduce round trips to the apiserver, a local unbound TTL cache is used to proxy userInfo requests. To prevent stale secrets from polluting the api server, a garbage collection goroutine is started to periodically reap expired sessions.

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -27,28 +30,45 @@ func main() {
...
+	var clientset k8s.Interface
+	if config.EnableRBAC || config.EnableInClusterStorage {
+		var err error
+		clientset, err = kubernetes.GetClientSet()
 		if err != nil {
 			log.Fatalf("error getting kubernetes client: %v", err)
 		}
 	} else {
 		clientset = nil
 	}
```
Comment (joejulian) [inline cmd/main.go:42]: No need. `clientset` will be nil.
```suggestion
	}
```

File: internal/server.go [hunk 2]
```diff
@@ -367,6 +354,9 @@ func (s *Server) authRedirect(logger *logrus.Entry, w http.ResponseWriter, r *ht
 		scope = []string{oidc.ScopeOpenID, "profile", "email", "groups"}
 	}
 
+	// clear existing claims session
+	_ = s.userinfo.Clear(r, w)
```
Comment (joejulian) [inline internal/server.go:358]:
```suggestion
	s.userinfo.Clear(r, w)
```

File: internal/storage/gorilla.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+type GorillaUserInfoStore struct {
+	SessionStore sessions.Store
+	SessionName  string
+}
+
+func (c *GorillaUserInfoStore) Get(r *http.Request) (*v1alpha1.UserInfo, error) {
+	session, _ := c.SessionStore.Get(r, c.SessionName)
+	if session == nil {
+		return nil, v1alpha1.UserDataStoreError(fmt.Sprintf("could not get session: %s", c.SessionName))
+	} else if session.IsNew {
```
Comment (joejulian) [inline internal/storage/gorilla.go:23]:
```suggestion
	}
	if session.IsNew {
```

Comment (jr0d) [reply internal/storage/gorilla.go:25]: Ah yes, there was something else after else if at one point I thiink

File: internal/authentication/auth.go [hunk 4]
```diff
@@ -236,93 +225,51 @@ func csrfCookieDomain(r *http.Request) string {
...
-	for _, d := range config.CookieDomains {
-		if d.Match(p[0]) {
-			return true, d.Domain
+	if a.config != nil {
+		for _, d := range a.config.CookieDomains {
+			if d.Match(p[0]) {
+				return true, d.Domain
+			}
 		}
 	}
```
Comment (joejulian) [inline internal/authentication/auth.go:238]:
```suggestion
	if a.config == nil {
		return false, p[0]
	}
	for _, d := range a.config.CookieDomains {
		if d.Match(p[0]) {
			return true, d.Domain
		}
	}
```

File: internal/storage/cluster/cluster.go [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+func (cs *ClusterStorage) getSecretByClaim(claimsId string) (*corev1.Secret, error) {
+	secrets, err := cs.getSecrets()
+	if err != nil {
+		return nil, fmt.Errorf("error getting secret list: %w", err)
+	}
+
+	for _, s := range secrets.Items {
+		cid, ok := s.ObjectMeta.Labels[storage.ClaimsIDLabel]
+		if !ok {
+			logger.Errorf("found managed secret not containing claimID")
```
Comment (joejulian) [inline internal/storage/cluster/cluster.go:210]: should the user take action on this error, and is there enough info to do so?

File: internal/storage/cluster/gc.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+func (gc *GarbageCollector) Collect() {
+	gc.collectorRunning = true
+	defer func() {
+		gc.collectorRunning = false
+	}()
+	gc.logDebug("collector starting")
+	err := gc.cs.deleteExpiredSecrets()
+	if err != nil {
+		gc.logDebug("error deleting expired secrets")
```
Comment (joejulian) [inline internal/storage/cluster/gc.go:39]: Should this include the error?

**Review by joejulian — APPROVED** [review]
This looks great. Added a few nits but overall :100: .

File: cmd/main.go [hunk 7]
```diff
@@ -26,29 +31,46 @@ func main() {
...
+	} else {
+		clusterStorage := cluster.NewClusterStore(
+			clientset,
+			config.ClusterStoreNamespace,
+			string(config.Secret),
+			config.Lifetime,
+			time.Duration(config.ClusterStoreCacheTTL)*time.Second,
+			authenticator)
+
+		gc := cluster.NewGC(clusterStorage, time.Minute, false, true)
```
Comment (jkoelker) [inline cmd/main.go:67]:
```suggestion
		userInfoStore := cluster.NewClusterStore(
			clientset,
			config.ClusterStoreNamespace,
			string(config.Secret),
			config.Lifetime,
			time.Duration(config.ClusterStoreCacheTTL)*time.Second,
			authenticator)

		gc := cluster.NewGC(userInfoStore, time.Minute, false, true)

```

**Review by jkoelker — APPROVED** [review]
LGTM, modulo the suggestion to use the clusterstorage directly as `userInfoStore` so it gets passed into `NewServer`

File: internal/storage/cluster/cluster.go (hunk 5 above)
Comment (jr0d) [reply internal/storage/cluster/cluster.go:210]: adding the secret namespace/name to the logging output.

File: internal/storage/cluster/gc.go (hunk 6 above)
Comment (jr0d) [reply internal/storage/cluster/gc.go:39]: yes it should :)

File: cmd/main.go (hunk 7 above)
Comment (jr0d) [reply cmd/main.go]: great catch! I found this while running e2e tests in konvoy, and somehow missed this comment :)
